### PR TITLE
convert : fix byte tokens for --vocab-type hfft

### LINKED ISSUE #5064: `convert.py --vocab-type hfft` produces `<0x0A>` instead of new lines
I am using llama.cpp master. I am trying to convert this model to GGUF: https://huggingface.co/tenyx/TenyxChat-8x7B-v1

After running `convert.py --vocab-type hfft`, the model will not output new lines correctly:

~~~
./main -m ~/TenyxChat-8x7B-v1-Q8_0.gguf -p "[INST]Write some poetry about typography.[/INST]" -n 128

system_info: n_threads = 8 / 16 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | 
sampling: 
        repeat_last_n = 64, repeat_penalty = 1.100, frequency_penalty = 0.000, presence_penalty = 0.000
        top_k = 40, tfs_z = 1.000, top_p = 0.950, min_p = 0.050, typical_p = 1.000, temp = 0.800
        mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000
sampling order: 
CFG -> Penalties -> top_k -> tfs_z -> typical_p -> top_p -> min_p -> temp 
generate: n_ctx = 512, n_batch = 512, n_predict = 128, n_keep = 0


 [INST]Write some poetry about typography.[/INST] In the world of words, where letters roam,<0x0A>Typography sets the tone, like a artist's home.<0x0A>Each stroke and curve, a deliberate choice,<0x0A>A symphony of shapes, with a unified voice.<0x0A><0x0A>Serif, sans-serif, or script,<0x0A>A message is sent, before even dict.<0x0A>In a font's design, we see the creator's mind,<0x0A>A reflection of culture, that's hard to find.<0x0A><0x0A>Some letters kiss, others stand apart,<0x0A>Creating rhythm and flow, with an artful heart.<0x0A>The negative space
llama_print_timings:        load time =    1578.15 ms
llama_print_timings:      sample time =      16.40 ms /   128 runs   (    0.13 ms per token,  7803.45 tokens per second)
llama_print_timings: prompt eval time =    1893.05 ms /    14 tokens (  135.22 ms per token,     7.40 tokens per second)
llama_print_timings:        eval time =   40938.67 ms /   127 runs   (  322.35 ms per token,     3.10 tokens per second)
llama_print_timings:       total time =   42878.28 ms /   141 tokens

~~~

Possibly related to #4622.

### PR DESCRIPTION
**Pull request by Artefact2**
This is inspired by 9f297f81adb93fdfefbd6496ff50cdfe070bc775, which got lost during the refactoring in 6efb8eb30e7025b168f3fda3ff83b9b386428ad6.

Fixes #5064.

### CODE REVIEW

File: convert.py [hunk 1]
```diff
@@ -509,11 +509,13 @@ def hf_tokens(self) -> Iterable[tuple[bytes, float, gguf.TokenType]]:
 
             # Convert token text to bytes
             token_text = reverse_vocab[token_id].encode("utf-8")
+            if token_text.startswith(b"<0x") and token_text.endswith(b">"):
```
Comment (cebtenzzre) [inline convert.py:512]: I prefer how the original explicitly checks that len(token) == 6. Even better, use a regex:
```python
if re.fullmatch(r'<0x[0-9A-F]{2}>', token_text):
```

Comment (cebtenzzre) [reply convert.py:512]: Even better, use a regex:
```suggestion
            if re.fullmatch(r'<0x[0-9A-F]{2}>', token_text):
```

**Comment by Artefact2** [comment]
Can I get a ack/nack on this?

**Comment by felladrin** [comment]
I've checked out your branch, compiled `main`, and tested it. It solved the issue for me 🙏 

Command:
```
llama.cpp/main -m model/ggml-model-f32.gguf -p "<|im_start|>user\nHow do I incorporate visual elements into my writing?<|im_end|>\n<|im_start|>assistant" -n 250 -e --repeat_penalty 1.0 -c 0 --temp 0.1 --min-p 0.1
```

Output (_without_ the changes from this PR):
```
<|im_start|>user<0x0A>How do I incorporate visual elements into my writing?<|im_end|><0x0A><|im_start|>assistant<0x0A>I do not have access to the specific details of the text material. However, here are some examples of how to use the following:<0x0A><0x0A>
```

Output (_with_ the changes from this PR):
```
<|im_start|>user
How do I incorporate visual elements into my writing?<|im_end|>
<|im_start|>assistant
I do not have access to the specific details of the text material. However, I can provide you with a list of the following:
```

So I hope it can be merged soon, as it will finally allow us to correctly convert Llama/Mistral models that doesn't have the `tokenizer.model` file along with the model.

**Comment by x4080** [comment]
waiting for this too

File: convert.py [hunk 2]
```diff
@@ -509,11 +509,13 @@ def hf_tokens(self) -> Iterable[tuple[bytes, float, gguf.TokenType]]:
 
             # Convert token text to bytes
             token_text = reverse_vocab[token_id].encode("utf-8")
+            if re.fullmatch(br"<0x[0-9A-Fa-f]{2}>", token_text):
+                toktype = gguf.TokenType.BYTE
+            else:
+                toktype = self.get_token_type(token_id, self.special_ids)
```
Comment (slaren) [inline convert.py:515]: Is there any reason to not implement this logic directly in `get_token_type`? It seems out of place here.

**Comment by ggerganov** [comment]
@Artefact2 Sorry for the delay - I'm somehow just seeing the PR

I see this is now closed - are you considering a new PR?

**Comment by Artefact2** [comment]
No, the bug is fixed as far as I'm concerned. If my patch is not up to the standard required, someone else can take care of it.

**Comment by maziyarpanahi** [comment]
> the bug is fixed

Thanks @Artefact2 for your contribution. Was it fixed by other PRs already merged into the main, or it was fixed for you and you moved on? (totally understand if it's the latter)

**Comment by Artefact2** [comment]
> Was it fixed by other PRs already merged into the main

Yes! See #5341.

**Comment by maziyarpanahi** [comment]
> > Was it fixed by other PRs already merged into the main
> 
> Yes! See #5341.

Many thanks!

**Pull request closed without merging**